### PR TITLE
handbook page explaing policy that all new internal APIs must be dervied from protobuf structs

### PR DESCRIPTION
**Pull request by ggilmore**
Closes https://github.com/sourcegraph/sourcegraph/issues/51066

Planning to post this in #dev-announce: 

> As we standardize on using gRPC for all of our internal APIs, there will be in an interim period where we have to provide both the gRPC and REST implementations of the same API. In order to ease our maintenance burden until we can safety delete the REST versions of our APIs, we're asking everyone to do the following:
> 
> New internal APIs must have both gRPC and REST implementations to reduce the maintenance burden. The REST implementation should use the [canonical JSON representation of the generated protobuf structs](https://protobuf.dev/programming-guides/proto3/#json) in the HTTP body for arguments and responses.

This handbook page will be linked in the announcement. Any feedback?

## Test plan
CI

### CODE REVIEW

File: doc/dev/background-information/gRPC.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Overview 
+
+New internal APIs must have both gRPC and REST implementations to reduce the maintenance burden. The REST implementation should use the c[canonical JSON representation of the generated protobuf structs](https://protobuf.dev/programming-guides/proto3/#json) in the HTTP body for arguments and responses.
```
Comment (camdencheek) [inline doc/dev/background-information/gRPC.md:3]:
```suggestion
New internal APIs must have both gRPC and REST implementations so we can provide a grace period for customers. The REST implementation should use the [canonical JSON representation of the generated protobuf structs](https://protobuf.dev/programming-guides/proto3/#json) in the HTTP body for arguments and responses.
```

File: doc/dev/background-information/gRPC.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+
+
+## simple example 
+
+The following example demonstrates how to implement a simple service in Go that provides both gRPC and REST APIs, using the  [canonical JSON representation of the generated Protobuf structs](https://protobuf.dev/programming-guides/proto3/#json).
```
Comment (camdencheek) [inline doc/dev/background-information/gRPC.md:9]:
```suggestion
The following example demonstrates how to implement a simple service in Go that provides both gRPC and REST APIs, using the [canonical JSON representation of the generated Protobuf structs](https://protobuf.dev/programming-guides/proto3/#json). 
```

File: doc/dev/background-information/gRPC.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+New internal APIs must have both gRPC and REST implementations to reduce the maintenance burden. The REST implementation should use the c[canonical JSON representation of the generated protobuf structs](https://protobuf.dev/programming-guides/proto3/#json) in the HTTP body for arguments and responses.
+
+
+
+## simple example 
+
+The following example demonstrates how to implement a simple service in Go that provides both gRPC and REST APIs, using the  [canonical JSON representation of the generated Protobuf structs](https://protobuf.dev/programming-guides/proto3/#json). 
+
+Note that the Go service uses [google.golang.org/protobuf/encoding/protojson](https://google.golang.org/protobuf/encoding/protojson) to Marshall and Unmarshall Protobuf structs to/from JSON. The standard "encoding/json" package should **not** be used here: it doesn't correctly operate on protobuf structs.
```
Comment (camdencheek) [inline doc/dev/background-information/gRPC.md:11]:
```suggestion
Note that the Go service uses [google.golang.org/protobuf/encoding/protojson](https://google.golang.org/protobuf/encoding/protojson) to Marshal and Unmarshal Protobuf structs to/from JSON. The standard "encoding/json" package should **not** be used here: it doesn't correctly operate on protobuf structs. 
```

**Review by camdencheek — APPROVED** [review]

File: doc/dev/background-information/gRPC.md (hunk 3 above)
Comment (ggilmore) [reply doc/dev/background-information/gRPC.md]: Yeah, there are some specifics for how to handle certain types in https://protobuf.dev/programming-guides/proto3/#json

timestamps are a concrete example:

```proto
syntax = "proto3";

import "google/protobuf/timestamp.proto";

message Example {
  string name = 1;
  int32 age = 2;
  google.protobuf.Timestamp created_at = 3;
}

```

If we have this struct:

```go
example := &examplepb.Example{
		Name:      "John Doe",
		Age:       30,
		CreatedAt: timestamppb.New(time.Now()),
	}

```

We could see a difference in the output (note that the spec mandates RFC-3339 formatted timestamps):


```jsonc
encoding/json output: {"name":"John Doe","age":30,"createdAt":{"seconds":1646841234,"nanos":123456789}}
protojson output: {"name":"John Doe","age":30,"createdAt":"2022-03-09T12:34:56.123456789Z"}
```

File: doc/dev/background-information/gRPC.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
+# Overview
+
+New internal APIs must have both gRPC and REST implementations so that we can provide a grace period for customers. The
+REST implementation should use
+the [canonical JSON representation of the generated protobuf structs](https://protobuf.dev/programming-guides/proto3/#json)
+in the HTTP body for arguments and responses.
+
+We expect to only have to maintain both implementations for the 5.1.X release in June. Afterwards, only the gRPC API will be used and we can delete the redundant REST implementations.
```
Comment (mucles) [inline doc/dev/background-information/gRPC.md:8]: Would it make sense to also note that there are some internal apis that don't make sense to migrate to gRPC so there would still be some left over REST implementations to maintain ?

**Review by mucles — APPROVED** [review]

File: doc/dev/background-information/gRPC.md (hunk 4 above)
Comment (ggilmore) [reply doc/dev/background-information/gRPC.md]: Good point! I modified the first section to mention this.

**Pull request merged**